ipam/manager_test.go

```diff
@@ -141,233 +146,279 @@ func cleanupTestAddressSpace(am AddressManager) error {
+					}
+					as.Pools["ap-test"] = ap
+					am.AddrSpaces["as-test"] = as
+					err := am.restore()
+					Expect(err).To(BeNil())
+					as = am.AddrSpaces["as-test"]
+					ap = as.Pools["ap-test"]
+					ar := ap.addrsByID["ar-test"]
+					Expect(ar.ID).To(Equal("ar-test"))
+					Expect(ap.RefCount).To(Equal(1))
+					Expect(ar.InUse).To(BeTrue())
+				})
+			})
+
+			Context("When GetModificationTime return error", func() {
+				It("Should not clear the RefCount and InUse", func() {
+					am := &addressManager{
+						AddrSpaces: make(map[string]*addressSpace),
+					}
+					am.store = &testutils.KeyValueStoreMock{
+						GetModificationTimeError: errors.New("Error"),
+					}
+					ap := &addressPool{
+						Id:			"ap-test",
+						RefCount: 	1,
+						Addresses: 	make(map[string]*addressRecord),
+					}
+					ap.Addresses["ar-test"] = &addressRecord{
+						ID:		"ar-test",
+						InUse: 	true,
+					}
+					as := &addressSpace{
+						Id:		"as-test",
+						Pools: 	make(map[string]*addressPool),
+					}
+					as.Pools["ap-test"] = ap
+					am.AddrSpaces["as-test"] = as
+					err := am.restore()
+					Expect(err).To(BeNil())
+					as = am.AddrSpaces["as-test"]
+					ap = as.Pools["ap-test"]
+					ar := ap.addrsByID["ar-test"]
+					Expect(ar.ID).To(Equal("ar-test"))
+					Expect(ap.RefCount).To(Equal(1))
+					Expect(ar.InUse).To(BeTrue())
+				})
+			})
+
+			Context("When rebooted", func() {
+				It("Should clear the RefCount and InUse", func() {
+					am := &addressManager{
+						AddrSpaces: make(map[string]*addressSpace),
+					}
+					am.store = &testutils.KeyValueStoreMock{}
+					ap := &addressPool{
+						Id:			"ap-test",
+						RefCount: 	1,
+						Addresses: 	make(map[string]*addressRecord),
+					}
+					ap.Addresses["ar-test"] = &addressRecord{
+						ID:		"ar-test",
+						InUse: 	true,
+					}
+					as := &addressSpace{
+						Id:		"as-test",
+						Pools: 	make(map[string]*addressPool),
+					}
+					as.Pools["ap-test"] = ap
+					am.AddrSpaces["as-test"] = as
+					err := am.restore()
+					Expect(err).To(BeNil())
+					as = am.AddrSpaces["as-test"]
+					ap = as.Pools["ap-test"]
+					ar := ap.addrsByID["ar-test"]
+					Expect(ar.ID).To(Equal("ar-test"))
+					Expect(ap.RefCount).To(Equal(0))
+					Expect(ar.InUse).To(BeFalse())
+				})
+			})
+		})
+
+		Describe("Test save", func() {
+			Context("When store is nill", func() {
+				It("Should return nil", func() {
+					am := &addressManager{}
+					err := am.save()
+					Expect(err).NotTo(HaveOccurred())
+				})
+			})
+		})
+
+		Describe("Test StartSource", func() {
+			Context("When environment is azure", func() {
+				It("Should return azure source", func() {
+					am := &addressManager{}
+					options := map[string]interface{}{}
+					options[common.OptEnvironment] = common.OptEnvironmentAzure
+					err := am.StartSource(options)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(am.source).NotTo(BeNil())
+				})
+			})
+
+			Context("When environment is mas", func() {
```

Comment:
It doesn't look like we're testing address manager functions anymore?
e.g. RequestAddress, ReleaseAddress, ...

Reply:
I found that functions such as RequestAddress just call the functions in pool.go and raise errors, there is no complicated work in itself. So I didn't test them. And the unit tests of functions in pool.go were all covered in pool_test.go in next PR :)